### PR TITLE
7 use thread safe date package

### LINKED ISSUE #7: Use thread safe date package
java.text.SimpleDateFormat is not thread safe and consequently code utilizes synchronized blocks or ThreadLocal .e.g.

These should be replaced with either JodaTime or DateTimeFormatter.

### PR DESCRIPTION
**Pull request by rdhayes68**

### CODE REVIEW

**Review by hlgp — APPROVED** [review]

**Review by mineralntl — APPROVED** [review]

File: src/main/java/datawave/util/time/DateHelper.java [hunk 1]
```diff
@@ -156,10 +203,15 @@ public static Date parseHour(String date) {
...
-        } else {
-            return Date.from(LocalDate.parse(lenientDate, parser).atStartOfDay(parser.getZone()).toInstant());
+        try {
+            if (hasTime) {
+                return Date.from(ZonedDateTime.parse(lenientDate, parser).toInstant());
+            } else {
+                return Date.from(LocalDate.parse(lenientDate, parser).atStartOfDay(parser.getZone()).toInstant());
+            }
+        } catch (DateTimeParseException e) {
+            log.warn(e.getMessage());
```
Comment (jschmidt10) [inline src/main/java/datawave/util/time/DateHelper.java]: Logging and then throwing will typically lead to multiple copies of this exception being logged. It's better to log and handle it OR just throw it and let whoever catches log it. In this case, just throwing it makes sense to me.

File: src/main/java/datawave/util/time/DateHelper.java [hunk 2]
```diff
@@ -213,7 +265,51 @@ public static String format8601(Date date) {
...
+        if ("yyyyDDD".equals(pattern) && "yyyyDDD".length() > date.length()) {
+            pattern = "yyyy D";
+            date = (date.length() > 5) ? new StringBuilder(date).insert(4, " ").toString() : date;
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern).withZone(ZoneOffset.UTC);
+            return Date.from(LocalDate.parse(date, formatter).atStartOfDay(formatter.getZone()).toInstant());
+        }
+        
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern).withZone(ZoneOffset.UTC);
+        
+        return lenientParseHelper(date, formatter, pattern, pattern.matches(hourRegex));
```
Comment (jschmidt10) [inline src/main/java/datawave/util/time/DateHelper.java]: Not sure how often we call this but if it's a non-trivial amount of times (like once per event during ingest/query), we could get a decent savings by making a constant for the compiled `hourRegex`. I believe the `matches()` call will have to recompile it every time.

Comment (rdhayes68) [reply src/main/java/datawave/util/time/DateHelper.java]: Done.

File: src/main/java/datawave/util/time/DateHelper.java (hunk 1 above)
Comment (rdhayes68) [reply src/main/java/datawave/util/time/DateHelper.java]: Done

**Review by hlgp — APPROVED** [review]

File: src/main/java/datawave/util/time/DateHelper.java [hunk 3]
```diff
@@ -213,7 +269,49 @@ public static String format8601(Date date) {
...
+        if ("yyyyDDD".equals(pattern) && "yyyyDDD".length() > date.length()) {
+            pattern = "yyyy D";
+            date = (date.length() > 5) ? new StringBuilder(date).insert(4, " ").toString() : date;
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern).withZone(ZoneOffset.UTC);
+            return Date.from(LocalDate.parse(date, formatter).atStartOfDay(formatter.getZone()).toInstant());
+        }
+        
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(pattern).withZone(ZoneOffset.UTC);
+        
+        return lenientParseHelper(date, formatter, pattern, hourMatcher.reset(pattern).matches());
```
Comment (jschmidt10) [inline src/main/java/datawave/util/time/DateHelper.java]: Sorry for the confusion. I actually think reusing matcher will make this not threadsafe again. You can build the Pattern statically and then create a new matcher on each invocation. Like:

```
public final static Pattern HOUR_PATTERN = Pattern.compile("...");

public static Date parseCustom(...) {
  boolean hasHours = HOUR_PATTERN.matcher(pattern).matches();
}
```

Comment (rdhayes68) [reply src/main/java/datawave/util/time/DateHelper.java]: Done.

**Review by jschmidt10 — APPROVED** [review]

File: src/main/java/datawave/util/time/DateHelper.java [hunk 4]
```diff
@@ -213,7 +268,49 @@ public static String format8601(Date date) {
      * @return the {@code Date} object
      */
     public static Date parse8601(String date) {
-        return Date.from(ZonedDateTime.parse(date, DTF_8601).toInstant());
+        try {
+            return Date.from(ZonedDateTime.parse(date, DTF_8601).toInstant());
+        } catch (DateTimeParseException e) {
+            log.warn(e.getMessage());
```
Comment (jschmidt10) [inline src/main/java/datawave/util/time/DateHelper.java]:
```suggestion
```

**Review by jschmidt10 — APPROVED** [review]

**Comment by jschmidt10** [comment]
Caught that other warn error. Not a huge deal but if you make another round of changes, I'd include it.

File: src/main/java/datawave/util/time/DateHelper.java (hunk 4 above)
Comment (rdhayes68) [reply src/main/java/datawave/util/time/DateHelper.java]: Done. Good catch.

**Review by jschmidt10 — APPROVED** [review]

**Review by hlgp — APPROVED** [review]

**Pull request merged**